ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +377,296 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
+      Map<Integer, Integer> groupIdIterationCountMap, MutableBoolean allReplicasCaughtUpEarly,
+      MutableBoolean maxIterationsReached) {
+
+    //check if thread is in shutdown or cycle termination is received
+    if (!running || shouldTerminateCurrentCycle) {
+      logger.trace(
+          "Thread name: {} not creating new groups as thread is in shutdown mode {} or terminate cycle received {}",
+          threadName, !running, shouldTerminateCurrentCycle);
+      return false;
+    }
+
+    //get max iterations by any remote replica group
+    int maxIterationCount = 0;
+    for (Map.Entry<Integer, Integer> entry : groupIdIterationCountMap.entrySet()) {
+      maxIterationCount = Integer.max(maxIterationCount, entry.getValue());
+    }
+
+    //check if any replica reached max iteration count
+    if (maxIterationCount == maxIterationsPerGroupPerCycle) {
+      if (maxIterationsReached.isFalse()) {
+        maxIterationsReached.setTrue();
+        if (inflightRemoteReplicaGroup.isEmpty()) {
+          allReplicasCaughtUpEarly.setTrue();
+          logger.trace("Thread name: {} All replicas marked to caught up early", threadName);
+        }
+      }
+      logger.trace("Thread name: {} not creating new groups as iteration limit {} reached", threadName,
+          maxIterationsPerGroupPerCycle);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Generates active and stand by replicas with no progress,
+   * checks and filters replicas which are throttled and are present in any iteration group
+   * @param activeReplicas {@link List}
+   * @param standbyReplicasWithNoProgress {@link List}
+   * @param remoteReplicas {@link List}
+   * @param remoteReplicaToThrottledTill {@link Map}
+   * @param standbyReplicasNoProgressReplicaGroupId {@link Integer}
+   * @param groupId {@link Integer}
+   * @param inflightRemoteReplicaGroup {@link Map}
+   */
+  private void getActiveAndStandByNoProgressReplicas(List<RemoteReplicaInfo> activeReplicas,
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress, List<RemoteReplicaInfo> remoteReplicas,
+      Map<RemoteReplicaInfo, Long> remoteReplicaToThrottledTill, Integer standbyReplicasNoProgressReplicaGroupId,
+      Integer groupId, Map<Integer, RemoteReplicaGroup> inflightRemoteReplicaGroup) {
+    //this will contain all replicas which are available for iteration
+    Set<RemoteReplicaInfo> nonIteratingNonPendingRemoteReplicas = new LinkedHashSet<>(remoteReplicas);
+
+    //remove all throttled replicas
+    remoteReplicas.forEach(remoteReplicaInfo -> {
+      if (remoteReplicaToThrottledTill.containsKey(remoteReplicaInfo)
+          && remoteReplicaToThrottledTill.get(remoteReplicaInfo) > time.milliseconds()) {
+        nonIteratingNonPendingRemoteReplicas.remove(remoteReplicaInfo);
+        throttleCount.inc();
+      }
+    });
+
+    //remove all replicas of corresponding group id if group is in iteration
+    if (inflightRemoteReplicaGroup.containsKey(groupId)) {
```

Comment:
this can be written better, if you use an abstraction like GroupId. I am guessing the intent here is if this group has replicas that are in the middle of replication, then we don't include them in the next set of RemoteReplicaGroups right ?